File: mobile/ios/Hearth/Native/UI/Rich.swift
import SwiftUI

private let tagPattern = try! NSRegularExpression(pattern: #"<(\w+)>(.*?)</\1>"#)

/// The web's `rich()` for the few strings that carry markup (`<b>name</b>`, `<r>A</r>`): each tag's
/// text bold, and in the colour given for its tag, if any. A tag named in `links` becomes a link
/// instead, which the system opens in the browser; the link carries nothing from the app.
/// Translations keep the same tags (the i18n test checks), so this never shows a raw tag.
func rich(_ text: String, colors: [String: Color] = [:], links: [String: URL] = [:]) -> AttributedString {
    let ns = text as NSString
    var out = AttributedString()
    var at = 0
    for m in tagPattern.matches(in: text, range: NSRange(location: 0, length: ns.length)) {
        out += AttributedString(ns.substring(with: NSRange(location: at, length: m.range.location - at)))
        var tagged = AttributedString(ns.substring(with: m.range(at: 2)))
        let tag = ns.substring(with: m.range(at: 1))
        if let url = links[tag] {
            tagged.link = url
        } else {
            tagged.inlinePresentationIntent = .stronglyEmphasized
            if let color = colors[tag] { tagged.foregroundColor = color }
        }
        out += tagged
        at = m.range.location + m.range.length
    }
    out += AttributedString(ns.substring(from: at))
    return out
}

File: mobile/android/app/src/main/kotlin/com/hearth/ui/Rich.kt
package com.hearth.ui

import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.LinkAnnotation
import androidx.compose.ui.text.TextLinkStyles
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.text.withLink
import androidx.compose.ui.text.withStyle

private val TAG = Regex("<(\\w+)>(.*?)</\\1>")

/**
 * The web's `rich()` for the few strings that carry markup (`<b>name</b>`, `<r>A</r>`): each tag's
 * text in the style given for it, bold when none is given. Translations keep the same tags (the
 * i18n test checks), so this never shows a raw tag.
 */
fun rich(text: String, styles: Map<String, SpanStyle> = emptyMap(), links: Map<String, String> = emptyMap()): AnnotatedString = buildAnnotatedString {
    var at = 0
    for (m in TAG.findAll(text)) {
        append(text.substring(at, m.range.first))
        val tag = m.groupValues[1]
        val url = links[tag]
        if (url != null) {
            // Opened by the system browser; the link carries nothing from the app.
            withLink(LinkAnnotation.Url(url, TextLinkStyles(SpanStyle(textDecoration = TextDecoration.Underline)))) { append(m.groupValues[2]) }
        } else {
            withStyle(styles[tag] ?: SpanStyle(fontWeight = FontWeight.Bold)) { append(m.groupValues[2]) }
        }
        at = m.range.last + 1
    }
    append(text.substring(at))
}
